### DISCUSSION TITLE
267

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#538] [Comments]
http://jojoldu.tistory.com/267

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

앗 고맙습니다! ^^

**Comment by jojoldu** [comment]
작성자: YUAMING

일 그만두고 전체적인 정리가 필요했는데, 잘 보고 갑니다. 감사합니다:)

**Comment by jojoldu** [comment]
작성자: 제리루즈

덕분에 무정지 배포까지 완료했습니다ㅎㅎ
근데 한가지 문제가 잇는데요
배포를해도 set1에서 set2로 스위칭이 안되길래 뭐가 문제일까 봤더니
~/app/nonstop/deploy.sh 파일 실행하면
cp: cannot create regular file '/home/ec2-user/app/nonstop/jar/xxxxxxxxxx.jar' : Permission denied 
복사할때 권한이 없다고 뜨고
kill -15 pid 할때도 아래와 같이
/home/ec2-user/app/nonstop/deploy.sh: line 45: kill: (592) - Operation not permitted
오류가 뜨더라구요
음 처음에 따라할때 직접 스위칭 할때는 잘됐던거 같은데 모든 작업완료 후에 직접 빌드/배포를 실행해보니까
안되는데 따로 권한을 줘야 되나요?

**Comment by orgin1127** [comment]
안녕하세요 따라오다가 외부 디렉토리 만들고 real-application.yml 만드는 파트에서 헷갈리는게 있어서 질문 드립니다.
외부에 디렉토리들을 만든다고 하셨는데 윈도우 기준으로는 어디쯤에 해당하는지 잘 모르겠습니다.  
완전 외부에 만들었다가 아무래도 아닌것 같아서 프로젝트 폴더 바로 밑에 폴더를 생성, yml을 만들고 .gitIgnore로 해당 폴더를 등록했습니다. 
InteliJ 사용자가 아니고 STS에서 실습중이라 Debug configuration 에서 Spring boot app 항목으로 set1을 테스트 시켜보니 화면에 set1 이라는 표시도 뜨고 접속도 가능합니다.
이 상태로 commit, push 하고 ec2에 접속해서 가장 상위에서 폴더들을 만들고 real-application.yml도 만들어 줬습니다. 
그 후 주소로 접속하니 갑자기 사이트가 엄청나게 느려졌고 (main의 경우는 몇분 걸려 뜨긴 하지만 modal등이 다 깨진 상태로 표시됩니다.) /profile의 경우는 사이트 응답이 너무 느려 표시가 불가능하다는 브라우저 메세지가 나옵니다.  
아무래도 설정의 오류가 있는것 같은데 외부 디렉토리를 윈도우 로컬 환경과 ec2에서는 어디에 만들면 될까요? 기초적인 질문이라 죄송합니다 ㅠㅠㅠㅠ

**Comment by orgin1127** [comment]
위 질문은 자체 해결했습니다. 설정 파일 꼬인거 풀고 yml 다시 만들고 하면서 천천히 하니 되더라구요
배포 스크립트 작성 과정에서 문제가 생겨서 질문드리려고 하는데요
배포스크립트 58행 response=$(curl -s http://localhost:$IDLE_PORT/health) 에서 접속이 안되서 response 변수에 아무것도 담기지 않아 계속 실패가 뜹니다.
-s 옵션을 제거하고 진행 상황을 보니 curl: (7) Failed to connect to localhost port 8889: Connection refused 메세지가 뜨는걸 발견하였습니다.
(저의 경우 set1 포트를 8889로, set2를 8887로, 기본 서버 포트는 8888로 셋팅하였습니다.)
health check이 10번 실패로 끝나고 난 뒤 
$ ps -ef|grep java 로 확인해보면
root     19286     1  0 07:16 ?        00:00:25 java -jar /home/ec2-user/app/travis/jar/*-0.0.4.jar
root     21695     1  9 11:36 pts/0    00:00:14 java -jar -Dspring.profiles.active=set1 /home/ec2-user/app/nonstop/jar/set1-*.jar
로 두 개가 올라간 것을 확인할 수 있었습니다.  혹시 조언 가능하시다면 조언 부탁 드리겠습니다. ㅠㅠ

**Comment by hongsii** [comment]
orgin1127 //
저도 actuator가 안되서 찾아보니까 "curl localhost:8080/actuator/health" 과 같이 써줘야하더라구요.
(참고 : https://spring.io/guides/gs/actuator-service/)

만약 디폴트 패스를 안쓰고 싶으시다면 management.endpoints.web.base-path=/ 라고 application.yml에 추가하시면 됩니다. 
(참고 : https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#production-ready-endpoints-custom-mapping)

**Comment by orgin1127** [comment]
답글 감사합니다. 저의 경우는 actuator는 잘 작동하여 /actuator를 붙이지 않고 /health만 했을때도 up이 뜨는것은 확인하였습니다. 어제 내내 붙잡고 이거저거 실험해본 결과 real-application.yml을 맨 처음 생성했을때 한 번만 curl 통신이 성공하여 nginx가 그 쪽을 바라보는걸 확인했습니다. 그 뒤 switch.sh를 사용하기 전에 다시 deploy.sh를 실행하면 위에 쓴 connection refused가 반복되더라구요. (이유를 진짜 모르겠습니다 ㅠㅠ 만들었을때 딱 한번만 작동해서 local -> set1 profile로 변경이 되고 그 뒤 set2로 바꾸면 java jar는 프로세스에 올라가나 nginx가 바라보질 못해 에러페이지 뜨는 이 상황..)
real-application.yml의 ec2내부 위치를 이리저리 다 바꿔봐도 소용이 없고.. 해결법이 감도 안와서 막막하네요..

**Comment by jojoldu** [comment]
@orgin1127 음.. 일단 윗 댓글을 보면

```
$ ps -ef|grep java 로 확인해보면
root 19286 1 0 07:16 ? 00:00:25 java -jar /home/ec2-user/app/travis/jar/-0.0.4.jar
root 21695 1 9 11:36 pts/0 00:00:14 java -jar -Dspring.profiles.active=set1 /home/ec2-user/app/nonstop/jar/set1-.jar
```

라고 하셨는데요
보시면 ```set1-.jar```로 되어있습니다.

제 블로그 과정을 따라가셨다면 ```set1-springboot-webservice.jar```가 되야하는데요
작성하신 deploy.sh의 코드에서 오타가 있는게 아닐까 추측됩니다!

스크립트내에 있는 각각의 변수들을 echo로 출력해서 확인해보셔야할것 같습니다!

**Comment by orgin1127** [comment]
늦은시간에 답글 감사합니다. 깜빡하고 설명을 적어놓지 않아서 죄송합니다. set1-뒤에 **를 썼는데 githup에서 자동으로 생략된것 같습니다.  실행시키면 set1- 뒤에 현재 제가 실습중인 프로젝트 명이 들어가는 점은 확인했습니다.

**Comment by jojoldu** [comment]
@orgin1127 아하! 그렇군요
흠 일단 localhost:8889로 요청하는건 Nginx를 거치지 않습니다
띄우신 스프링부트 set1 프로젝트를 다이렉트로 찌르는 요청이라서 Nginx를 거치지 않는데요
커넥션 리퓨즈는 해당 어플리케이션이 올라가지 않았을때 발생합니다
정상적으로 작동했는데, 주소가 잘못됐다면 404 에러가 발생해야 하구요

1) jar 파일 주변을 보면 nohup.out 파일이 있는데요 이 파일을 보시면 jar 실행 로그가 다 남아있는데요
이부분을 보시면 힌트를 얻을수있을것 같구요
2) 제가 봤을때는 지금 set1뜨고 set2가 정상적으로 안뜬것 같습니다. 지금 보여주신 java ps 는 디폴트, set1 만 떳구요
**원래는 set1, set2가 둘다** 있어야합니다.
set2가 정상적으로 올라가지 않았는데 여기를 띄우는 스크립트 부분을 디버깅해서 보셔야할것 같습니다

**Comment by orgin1127** [comment]
안된다고 이거저거 건드려서 꼬여있는게 많았는데 일단 풀고 이유를 찾은것 같습니다.
저는 현재 기본 포트 중복을 피하기 위해(8080은 다른 프로그램에서 사용중입니다.) application.yml에 
spring:
  profiles: 
    active: local 
  jpa: 
    properties: 
      hibernate: 
        dialect: org.hibernate.dialect.MySQL5InnoDBDialect 

---
spring:
  profiles: local 
  datasource: 
    data: classpath:data-h2.sql #
  jpa: 
    show-sql: true 
    hibernate: 
      ddl-auto: create-drop
  h2:
    console:
      enabled: true
server:
  port: 8888
로 설정했는데요
그냥 배포 시에는 8888이 잘 적용되지만 nonstop/jar안의 nohup을 확인해보니 set1 jar를 올릴 때 profile은 set1으로 잡히는데 실행 tomcat 포트가 기본인 8080으로 실행이 되어서 ec2 프로세스에는 올라가지만 health check반복문에 반응을 안하는 상태로 파악됩니다. 아무래도 real-application.yml을 인식 못하고 있는것 같은데.. 이 위치를 어디로 지정해줘야 하는지..
현재 main에는 다음과 같이 작성했습니다.
public static final String APPLICATION_LOCATIONS = "spring.config.location="
			+"classpath:application.yml,"
			+"app/config/terraceHouse/real-application.yml";

그리고 nonstop/jar 안의 nohup 로그는 다음과 같습니다.
2018-03-30 13:36:40.386  INFO 7379 --- [           main] c.S.Demo.SpringBootDemoApplication       : The following profiles are active: set1
2018-03-30 13:36:45.344  INFO 7379 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat initialized with port(s): 8080 (http)
2018-03-30 13:36:45.373  INFO 7379 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
(생략)
2018-03-30 13:36:52.178  INFO 7379 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2018-03-30 13:36:52.192  INFO 7379 --- [           main] c.S.Demo.SpringBootDemoApplication       : Started SpringBootDemoApplication in 12.968 seconds (JVM running for 13.89)

그리고 프로세스 상태입니다. 

root      6939     1  0 13:08 ?        00:00:18 java -jar /home/ec2-user/app/travis/jar/terraceHouse-0.0.6.jar
root      7379     1  3 13:36 pts/0    00:00:15 java -jar -Dspring.profiles.active=set1 /home/ec2-user/app/nonstop/jar/set1-terraceHouse.jar

**Comment by orgin1127** [comment]
추가적으로 이상한(?) 점은 real-application.yml을 처음 생성하고 deploy.sh를 실행하면 health  check에 성공합니다. 그래서 proxy_pass $service_url;로 바꾸고 설정하고 nginx 재시작하면 set1을 잘 바라보는데
그 뒤 다시 deploy.sh를 실행하면 real-application.yml의 설정을 무시하고 set2가 tomcat port 8080으로 배포되어서 health check에 실패합니다.  프로세스 상에는 올라가지만요

>현재 구동중인 set 확인
>set1
>application.jar 교체
> set2 에서 구동중인 APP pid확인
> 현재 구동중인 APP이 없으므로 종료하지 않음
> set2 배포
> set2 10초 후 health check 시작
> curl http://localhost:8887/health
nohup: appending output to ‘nohup.out’
curl: (7) Failed to connect to localhost port 8887: Connection refused

$ ps -ef | grep java
root      8207     1  0 14:19 ?        00:00:19 java -jar /home/ec2-user/app/travis/jar/terraceHouse-0.0.6.jar
root      8848     1  1 14:46 pts/0    00:00:17 java -jar -Dspring.profiles.active=set1 /home/ec2-user/app/nonstop/jar/set1-terraceHouse.jar
root      9330     1  7 15:05 pts/0    00:00:15 java -jar -Dspring.profiles.active=set2 /home/ec2-user/app/nonstop/jar/set2-terraceHouse.jar

$ curl -s localhost/profile
set1

혹시라도 감이 잡히시는 부분이 있으시다면 조언 부탁 드리겠습니다...

**Comment by jojoldu** [comment]
@orgin1127 음 set1이 정상적으로 작동되었다면 real-application.yml는 잘 불러오는거라 보시면 됩니다.
그럼 문제는 set2 설정값이 이상할것 같은데요
real-application.yml 코드에서 디비접속 정보를 가리고 캡쳐나 코드로 보여주실수있으신가요?

**Comment by orgin1127** [comment]
@jojodu 휴일이신데 이른 시간부터 답글 감사드립니다. 
real-application.yml에는 
---
spring: 
  profiles: set1 
server: 
  port: 8889
--- 
spring: 
  profiles: set2 
server: 
  port: 8887
로 작성해놨구요 저 이외의 코드는 없었습니다. profiles 부분은 스페이스로 들여쓰기 했습니다. 탭이 너무 많이 나가서.. 혹시 이게 관계 있었을까요.

그리고 다음 단계를 보니 db 연결 작업 전에 main의 application.yml로 set1, set2를 이동시키시는걸 보고
# 운영 환경 
--- 
spring.profiles: set1


server: 
  port: 8889
  
--- 
spring.profiles: set2 

server: 
  port: 8887

로 application.yml에 작성해 서버상에서 시험해보니 잘 되더라구요 switch도 잘 작동하고 전부 연동시켜서 commit & push 하면 set1, set2로 서로 계속 변경이 잘 되는걸 확인했습니다. 이래도 괜찮은건지..;;
이러고 나니 문제는 다음 단계인 db 연동 과정에서 spring.profiles.include: real-db을 application.yml에 넣고 real-application.yml을 
spring:
  profiles: real-db
  datasource:
        url: 제 db주소:3306/webServiceTH
        username: id
        password: pw
        driver-class-name: org.mariadb.jdbc.Driver
로 설정 해주고 로컬에서 테스트하면 데이터가 db의 테이블 안으로 잘 들어가는걸 확인하고 출력까지 확인했습니다.
그런데 서버로 commit push 하면 페이지에 db테이블에 넣은 데이터가 안 뜨네요. 
curl -s localhost/profile 하니 real-db 라고 나오는게 아마도 profile이 set1 or set2, real-db 이렇게 되어야 할 것 같은데..  뭔가 real-db라는 profile이 덮어 씌워지는것 같습니다. include가 잘못된건지..
최종적으로 main의 application.yml은 이런 형태가 되었습니다.

spring:
  profiles: 
    active: local 
  jpa: 
    properties: 
      hibernate: 
        dialect: org.hibernate.dialect.MySQL5InnoDBDialect 

---
spring:
  profiles: local 
  datasource: 
    data: classpath:data-h2.sql #local db setting
  jpa: 
    show-sql: true 
    hibernate: 
      ddl-auto: create-drop
  h2:
    console:
      enabled: true
server:
  port: 8888
  
# 운영 환경 
--- 
spring.profiles: set1
spring.profiles.include: real-db

server: 
  port: 8889
  
--- 
spring.profiles: set2 
spring.profiles.include: real-db

server: 
  port: 8887

**Comment by jojoldu** [comment]
@orgin1127 넵 
```spring.profiles```는 
```
spring:
 profiles:
```
와 동일한 효과를 냅니다.

yml 특성상 들여쓰기에 따라 옵션이 다르게 먹힐수 있는데요
탭이 아닌 스페이스로 들여쓰기를 하셔서 제대로 profiles 옵션이 안먹힌것 같습니다

디비 문제는 방향이 잘못된것 같습니다
**어플리케이션 실행이 정상적으로 실행되고, DB 데이터가 조회가 안된다면**
```real-application.yml```은 정상적으로 불러온것이고, ```real-db.yml```이 안불려진것입니다.
profile은 real-application.ym에서 include했기 때문에 포함된것 뿐입니다
ec2 서버에 db yml이 정상적인 위치에 있는지 확인해보세요!

**Comment by orgin1127** [comment]
@jojoldu
답글 감사합니다. 다음 단계에서 real-application.yml의 내용을 수정해서 profile을 real-db로 가지도록 하라고 하셨는데 real-db.yml 이라는 파일을 따로 만들어야 하나요?

**Comment by jojoldu** [comment]
@orgin1127 아 아뇨 real-application.yml 내용을 real-db로 변경하신게 제대로 안된것 같다는 이야기입니다.
스프링부트가 정상적으로 로딩된건 classpath:applcation.yml은 잘 불러온거고, db 내용이 잘 안불려진것인데
로컬에서 됐다는건 ec2에 있는 real-application.yml을 db 코드로 변경한 뒤에 제대로 안불려지는것 같습니다.

**Comment by orgin1127** [comment]
@jojoldu
정말 친절한 답변 감사드립니다. ㅠㅠ 제가 너무 질문을 많이 드려서 죄송하네요..

제 현재 real-application.yml은 다음과 같습니다.
<img src=https://i.imgur.com/CCBTYPu.png>
위치는 ~/app/config/프로젝트명/real-application.yml 입니다.

set1, set2 둘 다 profile에서 real-db 라는 단어를 반환합니다.
<img src=https://i.imgur.com/RR8LWHk.png>

그리고 프로세스에는 set1, set2로 올라가 있는게 확인됩니다.
<img src=https://i.imgur.com/vzTQQ4z.png>

마지막으로 main의 application.yml 입니다.
<img src=https://i.imgur.com/x2ZSKMB.png>

현재 경로는
public static final String APPLICATION_LOCATIONS = "spring.config.location="
			+"classpath:application.yml,"
			+"app/config/terraceHouse/real-application.yml";
으로 지정해 놓았습니다.

**Comment by jojoldu** [comment]
@orgin1127 음 일단 ```app/config/terraceHouse/real-application.yml```에서 ```app/```은 상대경로입니다.
그래서 부트 실행위치가 어디냐에 따라 패스위치가 달라져서
절대경로로 잡으셔야합니다.
```app/config/terraceHouse/real-application.yml``` 를
```/app/config/terraceHouse/real-application.yml``` 로 한번 변경해서 배포해보시겠어요?
그리고 ec2 서버의 real-application.yml 위치가 ```/app/config/terraceHouse/real-application.yml```가 맞는지도 확인해보시겠어요?
real-application.yml 위치로 가서 ```pwd``` 명령어를 실행해보시면 절대경로가 확인 됩니다.

**Comment by orgin1127** [comment]
@jojoldu
감사합니다 ㅠㅠ 드디어 됐네요 pwd 명령어가 도움이 됐습니다. 리눅스에 대해 공부 많이 해야겠다는걸 느끼네요 정말 자세하게 도와주셔서 감사합니다. 휴일에 ㅠㅠ

**Comment by jhleon86** [comment]
너무 잘보고 있습니다^^
orgin1127 님처럼 저도 profile 조회시 real-db 가 반환되어서 디버깅을 해보니 

    @GetMapping("/profile")
    public String getProfile () {
        return Arrays.stream(env.getActiveProfiles()) 
                .findFirst()   
                .orElse("");
    }
.findFirst 로 찾아지는것이 real-db 이고, set1, set2 등은 두번째 인자로 되어있었습니다.

그래서 .skip(1)
                .findFirst()
이와같이 첫번째것을 스킵하여 가지고오게 하였습니다
뭔가 좋은 방법같지는 않네요 ㅎㅎ 저거때문에 test 도 막아놔야 했고요...

또하나는 배포를 기동시켜 보다보니 최초 수행이 후 재반복시 deploy.sh 에서 수행되는 로그에 permission denied 메세지가 자꾸 발생하였습니다. 
그래서 저는 sudo cp, sudo kill 이렇게 sudo 를 앞에 붙여 해결하였습니다. 
남은 두편도 잘 보겠습니다. 

정말 감사합니다.

**Comment by jojoldu** [comment]
@jhleon86 님 확인 감사합니다 :)
빠르게 글을 작성하다보니 코드 품질이 좋지 않았네요 ㅠ
real-db, test 등으로 set1, set2가 정상 출력 안되면 List의 contains 메소드를 사용해서 set1을 갖고 있는지, set2를 갖고 있는지 확인후 set1,set2를 반환하는 것도 좋을것 같습니다 :)

**Comment by deepweller** [comment]
안녕하세요. 글 너무 잘보고 있습니다.
궁금한 점이 하나 있는데 혹시 nginx.conf 파일에 두개의 server 블록을 둬서 두개의 서비스를 사용할 수 있을까요?

**Comment by jojoldu** [comment]
@deepweller 아 넵 가능합니다.
보통 작은 회사에서는 사내에서 사용하는 작은 규모의 서비스(관리자 서비스등)는 한 서버에 다 올리고 Nginx로 서로 다른 포트를 Proxy pass 하기도 합니다.

**Comment by deepweller** [comment]
답변 빨리 주셔서 감사합니다.
server 블록에 listen 에 있는 포트를 바꿔서 추가하니깐 되네요!
감사합니다!

**Comment by jojoldu** [comment]
@deepweller 넵! 수고하세요 :)

**Comment by deepweller** [comment]
안녕하세요! 보다가 또 궁금한게 하나 생겨서요...ㅎㅎ
set1과 set2 두개 서비스를 띄우는데 배포가 완료되서 switch가 되면
기존에 떠있던 서비스는 종료시키지 않아도 되나요?
감사합니당!

**Comment by jojoldu** [comment]
@deepweller 기존에 떠있던 서비스라는게 set2를 올린뒤 남아있던 set1을 말씀하시는거라면 저같은 경우엔 둘다 올려놓습니다.
이유는 빠른 롤백때문인데요.
배포했던  set2 서비스가 배포하고나니 버그가 있어서 장애가 난다면, 바로 set1로 스위칭을 하면 수초내로 서비스를 정상화 시킬 수 있습니다.
하지만 set1이 꺼져있다면 빠르게 롤백이 안됩니다.
그래서 저같은 경우엔 둘다 켜놓습니다.

**Comment by deepweller** [comment]
아아 그렇군요! 
그럼 물론 지금 진행하고 있는 이 서비스는 규모가 작지만 실제 서비스하는 경우에는 클 수가 있을텐데
두 개를 서버에서 실행하고 있으면 서버자원을 그만큼 더 차지하고 있어서 생기는 문제는 없을까요??
또 이러한 서비스가 여러개라면 각각 배포를 위해서 두개씩 실행되고 있을 수도 있을 것 같은데 이런 경우에도 상관이 없을까요??
서버의 성능에 따라 다르겠지만 회사에서 개인프로젝트로 스프링부트기반 서비스를 만들어서 이번장에서 설명해주신것 처럼 했을때 위와같은 질문을 받을 수도 있을 것 같아서요...ㅎㅎ

**Comment by jojoldu** [comment]
아! 저 방식은 기존에 일일 2천만 PV (페이지뷰) 발생하는 서비스에서도 동일하게 썼던 방식입니다.
2개를 띄우는 서버 자원은 거의 차이가 없을거에요.
서버 자원이라는게 결국 리퀘스트가 발생해야만 높아지지 부트 2개를 실행시키는걸로 크게 사양 차이는 없습니다.
(당연히 케이스바이 케이스로 어떻게 서비스를 구현했느냐에 따라 완전히 다르긴 합니다.
켜져있기만해도 자동으로 뭔가 실행되는 구조라면 당연히 자원이 계속 소모됩니다.)

다만. DB 커넥션풀이 부트 서비스 2개(set1, set2)가 사용하기 때문에 절반 밖에 못쓰는게 문제일것 같습니다.
DB 커넥션풀을 온전히 다 쓰시려면 위 방식의 무중단 배포는 못씁니다.
서버 2대 (여기선 EC2가 되겠네요) 가 필요하기 때문에 말씀하신 작은 서비스 기준은 안될것 같습니다.
(커넥션 풀은 서비스 실행 시점에 잡히지 자원이 남았다고 갑자기 증가하거나하지 않습니다.)

한 서버에 2대를 띄우는 방식이 아닌 무중단 배포를 하실려면
AWS에서는 ELB + EC2대를 사용하시면 됩니다.

**Comment by deepweller** [comment]
자세한 답변 감사합니다!!
덕분에 정말 많이 배우고 있습니다.

**Comment by jojoldu** [comment]
네^^ 감사합니다 :)

**Comment by gonghojin** [comment]
오늘도 잘보고 갑니다. 좋은 주말 보내세요~

**Comment by Integerous** [comment]
덕분에 많이 배우고 있습니다! 감사합니다ㅎ 

그런데 다 잘되다가 본문 제일 마지막에서 처럼 `주소/`를 치면 404가 뜹니다. 

git 커밋, 푸시를 하고`주소/profile` 로 접속하면 set1과 set2가 번갈아가며 잘 나오는데 `주소/`로 접근하면 404가 뜨네요. 배포파일의 문제인가 싶어 8080에 배포하고 `주소:8080/`으로 접속하면 이상없이 잘 됩니다.

다만 한 가지 걸리는 점이 webRestController로 Get매핑되는 `/hello`나 '/profile`은 동작하지만, webController에서 매핑되는 것들은 동작하지 않습니다. RestController 와 일반 Controller의 차이 때문일까요? 꽤 오랫동안 이리저리 찾아보고 오타도 여러 번 검사해봤습니다만 답을 못찾고 있습니다. 힌트가 될만한 게 있을까요?

**Comment by choish1919** [comment]
안녕하세요 항상 잘 보고 있습니다! 프로필 변경이 안되서 질문드려요.

로컬 환경에서 테스트 결과는 포트와 프로필이 정상적으로 바뀌는 것을 확인했는데
ec2에서는 프로필이 바뀌는것은 확인했으나 포트가 8081로 바뀌지 않네요.
그래서 자꾸 health 체크도 8081 포트에 접근할 수 없다고만 나오고

nohup.out 로그입니다.
2018-11-07 06:42:29.683  INFO 31142 --- [           main] com.*.*.Application               : The following profiles are active: set1
2018-11-07 06:42:35.146  INFO 31142 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)

real-application.yml 파일 경로는 로컬은 윈도우 환경이라 경로가 달라서
git에 푸시할때는 /app/config/프로젝트명/real-application.yml로 바꾼뒤 커밋 했어요.
경로가 잘못됬나 싶어 /home/ec2-user/app/config/프로젝트명/real-application.yml
/app/config/프로젝트명/real-application.yml 두 경로에 모두 파일을 만들고 해봤으나
결과는 똑같습니다.

java 검색시 아래와 같이 set1 프로필로 jar가 실행되는 것은 확인할 수 있는데 real-application.yml 파일을 못읽어서 포트가 8081로 안바뀌는 것 같은데..
root     30580     1  0 06:02 ?        00:00:21 java -jar /home/ec2-user/app/travis/jar/프로젝트명-0.0.3.jar
root     31142 31134  0 06:42 pts/0    00:00:00 java -jar -Dspring.profiles.active=set1 /home/ec2-user/app/nonstop/jar/set1-프로젝트명.jar
root     31144 31134  0 06:42 pts/0    00:00:00 grep java

아 그리고 deploy.sh가 다 끝난 뒤에 검색하면 
root     30580     1  0 06:02 ?        00:00:21 java -jar /home/ec2-user/app/travis/jar/프로젝트명-0.0.3.jar
root     31144 31134  0 06:42 pts/0    00:00:00 grep java
set1버전 jar는 검색이 안되더라고요. 이게 문제일까요?

real-application.yml 내용은 아래와 같습니다
---
spring:
  profiles: set1
server:
  port: 8081

---
spring:
  profiles: set2
server:
  port: 8082

혹시 제가 잘못 진행하거나 잘못 알고 있는게 있을까요???

**Comment by inganyoyo** [comment]
세션이 끊어질것 같은데 세션을 유지?할수있는 방법은 어떤게 있을까요?

**Comment by Gunggo** [comment]
real-application 파일 경로로 고생하고있는데 프로젝트 내에 app폴더 만들고 ignore로 등록한 후에 똑같이 spring.config.location 지정하면 될까요 ?
file:./ 경로로 잡아서 하면 로컬에서만 되고 ec2에선 안읽히고
/app로 시작하게 해도 리눅스에서 못읽어오네요

**Comment by Vermouth-Singed** [comment]
7-3-3. 배포 스크립트 작성
에서 status up 이 아니라는데 어떤게 잘못된걸까요?

**Comment by Vermouth-Singed** [comment]
위에꺼 질문 지우려고 했는데 안지워지네요...다 잘되는거 확인해서 신난다~이러고 이제 파일 정리하다가 갑자기 안되서 멘붕이네요...일단 nonstop/deploy.sh 실행시켰는데 여기서 cp 에러나는데 단순히 권한문제가 아니라 파일명 같으면 생길수도 있는건가요?

**Comment by Vermouth-Singed** [comment]
UP 문자열이 안생기는건 왜 그런건가요?

**Comment by min9288** [comment]
혹시 블로그 예제 보시고 하는 분들중에 SecurityConfig 구성하신분은 "/health" 랑 "/profile" permitAll() 꼭 해주세요
너무 당연한 건데,,,해결한다고 시간을 얼마나 소비했는지..

**Comment by seeyacht** [comment]
소스 수정없이 테스트에서는 잘 되었는데 소스를 수정하고 jar 교체 시 스위칭 전까지 기존 서비스가 각 페이지 호출 시 에러(/profile은 잘 실행됨)가 발생하여 로그를 확인 해보니 타임리프에서 에러가 발생하였습니다. 원인을 확인 해보니 ln 으로 실행하는 심볼릭 링크가 문제 였습니다. 그래서 심볼릭 링크 대신 cp로 jar파일을 생성하였더니 그런 문제가 사라졌습니다. 저처럼 에러나시는분들에게 도움이 되었으면 좋겠습니다.

**Comment by kimgun95** [comment]
set1 빌드, 배포가 실행되는 것? 처럼 보이지만 페이지는 예전의 것을 띄우는 분들이 있다면 이런 상황일 수 있어요!!!
처음 nonstop/deploy.sh 파일을 작성할 때 프로젝트명 변경을 미처 못하고 한 번 실행을 하면 변경 못한 이름으로 8081 port가 실행되고 있을 겁니다! 그래서 후에 프로젝트명 변경 후 set1 profile로 빌드가 제대로 되어도 정작 8081 port로는 다른 프로젝트가 배포되고 있는 겁니다,,하하 이걸로 애 좀 먹었네요.
set2 profile은 제대로 접근이 되는 것 같은데 set1 profile은 안되니까,,, 무튼 누군가의 해결책이 되길 바랍니다:)
